src/main/java/io/cdap/plugin/http/source/common/http/HttpResponse.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.http.source.common.http;
+
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.util.EntityUtils;
+
+import java.io.ByteArrayInputStream;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * A response from an http endpoint.
+ */
+public class HttpResponse implements Closeable {
+  private CloseableHttpResponse response;
+  private byte[] bytes;
+  private String body;
+
+  public HttpResponse(CloseableHttpResponse response) {
+    this.response = response;
+  }
+
+  public int getStatusCode() {
+    return response.getStatusLine().getStatusCode();
+  }
+
+  public Header[] getAllHeaders() {
+    return response.getAllHeaders();
+  }
+
+  public Header getFirstHeader(String headerName) {
+    return response.getFirstHeader(headerName);
+  }
+
+  public InputStream getInputStream() throws IOException {
+    if (bytes != null) { // CloseableHttpResponse already read and closed the stream. So we need to use existing body.
+      return new ByteArrayInputStream(bytes);
+    } else {
+      HttpEntity responseEntity = response.getEntity();
+      return responseEntity.getContent();
+    }
+  }
+
+  public String getBody() {
```

Comment:
I don't think this is a clean API since it's expected that the caller will call either getInputStream() or getBody() or getBytes(). It's probably better for the Page classes just to get the CloseableHttpResponse like they were getting before, with each classes reading the response in the way that makes sense for it.

Reply:
CloseableHttpResponse closes the response when read. So it cannot be read second time. It can pose a problem sometimes.

That's why I implemented this. Do you still think we should remove this?

Reply:
This class does not protect the caller though. For example, if somebody calls getInputStream() and then getBody(), won't that result in an error? This class tries to hide the fact that you can't read the response twice, but there is no way to provide that abstraction without storing the entire contents.

Reply:
Generally getInputStream() is called only by BasePage children. Which is the last step of the process. Should we maybe throw an exception if we try to getBody() is called after getInputStream()?

The wrapper protects from other situations like "We read body first time to pass it into custom pagination python code as an argument and second time for text format to read it as stream."

I'm not sure how we can overcome situations like this (there are also the same situation with errors). Without saving the whole string in wrapper.

Reply:
Hm, there must be a better way to structure this, but I suppose this is fine for now. Please add javadocs so the expected usage is clear.